genesis/streaming.py

```diff
@@ -14,350 +14,414 @@
-			from tqdm.autonotebook import tqdm
-
-		t = tqdm(disable=not progress, total=limit, desc='Alerts processed', unit=' alerts', mininterval=0.5, smoothing=0.5, miniters=0)
-
-		nread = 0
-		for rec, meta in self._filtered_stream(mapper=mapper, filter=filter, timeout=timeout):
-			yield (rec, meta) if self._metadata else rec
-
-			t.update()
-			nread += 1
-
-			if nread == limit:
-				break
-
-		t.close()
-
-	# returns a generator for the user-mapped messages using the filter function,
-	# possibly executed on ncores and up to maxread values
-	def __call__(self, filter=None, pool=None, progress=False, timeout=None, limit=None):
-		if pool:
-			mapper = lambda fun, vec: pool.imap(fun, vec, chunksize=100)
-		else:
-			mapper = map
-
-		yield from self._stream(filter, mapper=mapper, progress=progress, timeout=timeout, limit=limit)
-
-	def __iter__(self):
-		return self.__call__()
-
-	# producer functionality
-	def write(self, msg):
-		packet = self._serialize(msg)
-		self.p.produce(self.topics[0], packet)
-
-	def flush(self):
-		return self.p.flush()
+    c = None
+    p = None
+
+    def __init__(
+        self,
+        broker_url,
+        mode="r",
+        start_at="latest",
+        format="avro",
+        metadata=False,
+        config=None,
+    ):
+        self.groupid, self.brokers, self.topics = parse_kafka_url(broker_url)
+
+        # mode can be 'r', 'w', or 'rw'; other characters are ignored
+        assert "r" in mode or "w" in mode
+
+        if self.groupid is None:
+            # if groupid hasn't been given, emulate a low-level consumer:
+            #   - generate a random groupid
+            #   - disable committing (so we never commit the random groupid)
+            #   - start reading from the earliest (smallest) offset
+            import getpass, random
+
+            self.groupid = (
+                getpass.getuser()
+                + "-"
+                + "".join(
+                    random.choice(string.ascii_uppercase + string.digits)
+                    for _ in range(20)
+                )
+            )
+        # 			print(f"Generated fake groupid {self.groupid}")
+
+        # load librdkafka configuration file, if given
+        # configurable properties: https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
+        cfg = dict()
+        if config is not None:
+            if isinstance(config, dict):
+                cfg = config
+            elif isinstance(config, str):
+                import builtins
+
+                with builtins.open(config) as fp:
+                    parser = configparser.ConfigParser()
+                    parser.read_string("[root]\n" + fp.read())
+                    cfg = dict(parser["root"])
+
+        if "r" in mode:
+            ccfg = {
+                **cfg,
+                "bootstrap.servers": self.brokers,
+                "group.id": self.groupid,
+                "default.topic.config": {"auto.offset.reset": start_at},
```

Comment:
Having the value dict on one line seems less readable; the old version made the fact the value is a nested dict more apparent. Keps the sort of JSON-y/YAML-y look when you're skimming through the file (and diffs are cleaner as new entries are added).